fix for #57598 (ROLLBACK should be no-op...)

PapaToemmsn opened the pull request:
Change to verify proper behaviour for ROLLBACK should be no-op when no active tranaction is running (MySQL compatibility) #57598

### Changelog category (leave one):
- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Issue #57598 mentions a variant behaviour regarding transaction handling. An issued COMMIT/ROLLBACK when no transaction is active is reported as an error contrary to MySQL behaviour.

This leads to breaking the MySQL compatibility layer as some tools rely on the MySQL way to handle transactions.

Changes were tested in an environment where Clickhouse was used as a MySQL replacement.

PapaToemmsn committed:
fix for #57598 (ROLLBACK should be no-op...)

Change to verify proper behaviour for ROLLBACK should be no-op when no active tranaction is running (MySQL compatibility) #57598

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=60338) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=60338) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=60338) it.</sub>

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #60338 from PapaToemmsn/ClickHouse:test-transaction-fixes -->
*This is an automated comment for commit ed4636b53c71492e17d6b3584e625e3bec6c3bc2 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60338/ed4636b53c71492e17d6b3584e625e3bec6c3bc2/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60338/ed4636b53c71492e17d6b3584e625e3bec6c3bc2/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

PapaToemmsn committed:
adjusted 01173_transaction_control_queries.sql to changed no transaction active bahaviour

Adapted transaction test queries to changed beahviour

PapaToemmsn commented:
Tried to search for information/docs on how to proceed as the results of the tests mostly seem unrelated to my changes done (e.g. "DB::Exception: io_uring is not supported by this system:" is a nerror message for quite a lot of the failed tests).

Since I was not successful in finding hints regarding these problems any hint will be greatly appreciated.

alexey-milovidov commented:
The main problem is:

> Bugfix validation — New test(s) failed to reproduce a bug 

I've changed it to "Imrovement".

alexey-milovidov commented:
> PR Check — expect adding docs for pr-feature 

It looks like it was also mislabeled as "New Feature"

alexey-milovidov commented on tests/queries/0_stateless/01173_transaction_control_queries.sql:
```diff
@@ -6,8 +6,8 @@ drop table if exists mt2;
 
-commit; -- { serverError INVALID_TRANSACTION } -- no transaction
-rollback; -- { serverError INVALID_TRANSACTION }
+commit;
```
It should apply only to the MySQL compatibility. I don't think this change is correct. We can ask @tavplubix

tavplubix replied:
I prefer the strict version (and for `rollback` as well)

PapaToemmsn replied:
The issue is that this way it breaks MySQL tools as cited in #57598 so that Clickhouse can't be used as a drop in replacement for MySQL.
I share your opinion that the strict version makes sense but this denies easy usage in such scenarios.

Is there a way to get the information that the query is coming from MySQL port from the session context? If so this behaviour could be set only for such queries. Woudl that be a satisfying solution for you?

tavplubix replied:
Yes, it makes sense, see `Context::getClientInfo()` and `ClientInfo::Interface`

PapaToemmsn replied:
@tavplubix 

> Yes, it makes sense, see `Context::getClientInfo()` and `ClientInfo::Interface`

Thank you for your indications. I have changed the code based on ClientInfo::Interface to show this behaviour only for MySQL interfaces. I reverted back the test and created a new one for the MySQL client to verify the new behaviour.

Is there anything else I need to do to make this pull request mergeable?

CI seems ok as the 6 tests that did not work are no related to my changes.

Thank you!

PapaToemmsn committed:
Reveted 01173_transaction_control_queries.sql back to original state

As changes in Txn handling for COMMIT and ROLLBACK should only be honoured for MySQL client mode the test was reverted back

PapaToemmsn committed:
COMMIT/ROLLBACK with no active transaction do not throw an exception for MySQL clients anymore

PapaToemmsn committed:
COMMIT/ROLLBACK with no active transaction do not throw an exception for MySQL clients anymore

PapaToemmsn committed:
COMMIT/ROLLBACK with no active transaction do not throw an exception for MySQL clients anymore

PapaToemmsn committed:
fix style check errors

PapaToemmsn committed:
made 02989_mysql_transaction_test.sh executable

PapaToemmsn committed:
Merge branch 'ClickHouse:master' into test-transaction-fixes

PapaToemmsn committed:
Merge branch 'ClickHouse:master' into test-transaction-fixes
